lib/selinux.c

```diff
@@ -98,6 +99,146 @@ int reset_selinux_file_context (void)
+ */
+__attribute__((__format__ (printf, 2, 3)))
+static int selinux_log_cb (int type, const char *fmt, ...) {
+	va_list ap;
+	char *buf;
+	int r;
+#ifdef WITH_AUDIT
+	static int selinux_audit_fd = -2;
+#endif
+
+	va_start (ap, fmt);
+	r = vasprintf (&buf, fmt, ap);
+	va_end (ap);
+
+	if (r < 0) {
+		free (buf);
+		return 0;
+	}
+
+#ifdef WITH_AUDIT
+	if (-2 == selinux_audit_fd) {
+		selinux_audit_fd = audit_open ();
+	}
+
+	if (selinux_audit_fd < 0) {
+		/* You get these only when the kernel doesn't have
+		 * audit compiled in. */
+		if (   (errno != EINVAL)
+		    && (errno != EPROTONOSUPPORT)
+		    && (errno != EAFNOSUPPORT)) {
+
+		    (void) fputs (_("Cannot open audit interface.\n"),
+		              stderr);
+		    SYSLOG ((LOG_WARN, "Cannot open audit interface."));
+		}
+	} else {
+		uid_t audit_uid = geteuid ();
+
+		if (SELINUX_AVC == type) {
+			audit_log_user_avc_message (selinux_audit_fd,
+			    AUDIT_USER_AVC, buf, NULL, NULL,
+			    NULL, audit_uid);
+			goto skip_syslog;
+		} else if (SELINUX_ERROR == type) {
+			audit_log_user_avc_message (selinux_audit_fd,
+			    AUDIT_USER_SELINUX_ERR, buf, NULL, NULL,
+			    NULL, audit_uid);
+			goto skip_syslog;
+		}
+	}
+#endif
+
+	SYSLOG ((LOG_WARN, "libselinux: %s", buf));
+
+skip_syslog:
+	free (buf);
+
+	return 0;
+}
+
+/*
+ * check_selinux_permit - Checks whether SELinux grants the given
+ *                        operation
+ *
+ *   Parameter is the SELinux permission name, e.g. rootok
+ *
+ *   Returns 0 when permission is granted
+ *                  or something failed but running in
+ *                  permissive mode
+ */
+int check_selinux_permit (const char *perm_name)
+{
+	return check_selinux_permit_shortcut (perm_name, NULL, 0);
+}
+
+/*
+ * check_selinux_permit - Checks whether SELinux grants the given
+ *                        operation
+ *
+ *   First parameter is the SELinux permission name, e.g. rootok
+ *   Second parameter if not NULL enables a shortcut for changes
+ *       on non-root and matching SELinux user.
+ *   Third parameter tells the shortcut what userid gets changed.
+ *       (if no shortcut is intented, pass 0)
+ *
+ *   Returns 0 when permission is granted
+ *                  or something failed but running in
+ *                  permissive mode
+ */
+int check_selinux_permit_shortcut (const char *perm_name,
+                                      const char *changed_user,
+                                      uid_t changed_uid)
+{
+	char *user_context_str;
+	int r;
+
+	if (0 == is_selinux_enabled ()) {
+		return 0;
+	}
+
+	selinux_set_callback (SELINUX_CB_LOG, (union selinux_callback) selinux_log_cb);
+
+	if (getprevcon (&user_context_str) != 0) {
+		fprintf (stderr,
+		    _("%s: can not get previous SELinux process context: %s\n"),
+		    Prog, strerror (errno));
+		SYSLOG ((LOG_WARN,
+		    "can not get previous SELinux process context: %s",
+		    strerror (errno)));
+		freecon (user_context_str);
```

Comment:
No need to free here; getprevcon() failed.  And user_context_str could be uninitialized.

Reply:
fixed, thanks